hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerSet.java

```diff
@@ -235,18 +235,20 @@ public void listContainer(long startContainerId, long count,
   public ContainerReportsProto getContainerReport() throws IOException {
     LOG.debug("Starting container report iteration.");
 
+    ContainerReportsProto.Builder crBuilder =
+        ContainerReportsProto.newBuilder();
     // No need for locking since containerMap is a ConcurrentSkipListMap
     // And we can never get the exact state since close might happen
     // after we iterate a point.
     List<Container<?>> containers = new ArrayList<>(containerMap.values());
-
-    ContainerReportsProto.Builder crBuilder =
-        ContainerReportsProto.newBuilder();
-
-    for (Container<?> container: containers) {
-      crBuilder.addReports(container.getContainerReport());
+    // Incremental Container reports can read stale container information
```

Comment:
Is there a potential race condition on line 243 where the containers array list is formed? The map is concurrent, so we probably cannot get a concurrentModification error, but could we get a snapshot of the containers here and meanwhile a new container is added and ICR queued that is not in our snapshot and hence never gets added to the FCR?

I think we need to synchronize around the forming of this list.

I also wonder why we need to form the new list? I guess we can release any lock faster if we are not forming the protobuf under the lock, but then if we synchronize here to block ICRs, I think we are effectively blocks new container creation anyway. Maybe we can avoid the new array list?

Need to think about this carefully.

Reply:
> but could we get a snapshot of the containers here and meanwhile a new container is added and ICR queued that is not in our snapshot and hence never gets added to the FCR?

Seeing how this is synchronized in StateContext, I don't think this could happen.

I do think we could end up with a scenario where the new container is added to the set, but the ICR is blocked from being queued due to the synchronization. Then we generate teh FCR that contains it, and then we also end up with a duplicate ICR. However I doubt that matters - on SCM it will just process it again, so it should be fine.